drivers/SmartThings/sonos/src/token_refresher.lua

```diff
@@ -0,0 +1,128 @@
+local cosock = require "cosock"
+local utils = require "utils"
+local security = require "st.security"
+local log = require "log"
+
+local module = {}
+
+local STATES = {
+  -- OAuth is not connected.
+  DISCONNECTED       = 1,
+  -- Driver has a valid token.
+  WAITING_FOR_EXPIRE = 2,
+  -- Driver either doesn't have a token or it is invalid.
+  REQUEST_TOKEN      = 3,
+}
+
+local STATE_STRINGIFY = {
+  [STATES.DISCONNECTED] = "disconnected",
+  [STATES.WAITING_FOR_EXPIRE] = "waiting for expire",
+  [STATES.REQUEST_TOKEN] = "requesting token",
+}
+
+local Refresher = {}
+Refresher.__index = Refresher
+
+--- Determine which state the refresher should be in.
+--- Any state can go to any state so this just depends on:
+--- - Is Oauth connected?
+--- - Do we have a valid token?
+function Refresher:determine_state()
+  if not self.driver:oauth_is_connected() then
+    -- Oauth is disconnected so no point in trying to request a token until we are connected.
+    return STATES.DISCONNECTED
+  end
+  local token, _ = self.driver:get_oauth_token()
+  if token then
+    local now = os.time()
+    local expiration = math.floor(token.expiresAt / 1000)
+    if (expiration - now) > 60 then
+      -- Token is valid and not expiring in the next 60 seconds.
+      return STATES.WAITING_FOR_EXPIRE
+    end
+  end
+  -- We don't have a valid token or it is about to expire soon.
+  return STATES.REQUEST_TOKEN
+end
+
+--- Waits for a token event with a timeout.
+function Refresher:receive_token(timeout)
+  local token_bus, _ = self.driver:oauth_token_event_subscribe()
+  if token_bus then
```

Comment:
So I think the non-existent `else` block of this if kind of breaks some assumptions I would have about how this function and `wait_for_expire` work.  That is for this function I would expect that after calling this function, we will have a valid token, but it is not clear to me what state we would/should be in if token_bus is falsey.